design/design.md

```diff
@@ -0,0 +1,251 @@
+implemented in the future; for instance, unweighted variants can be concretely
+more efficient due to no lottery and less complexity, and in the decentralized
+setting, an unweighted version could be more efficient due to not needing to
+sample the binomial distribution.
+
+We should be able to reuse most of the logic between different schemes anyway,
+so the overhead shouldn't be large.
+
+## Choice of hash function
+The "random" functions $H_0$, $H_1$, $H_2$ and others in ALBA constructions need to be
+implemented using some concrete function, either a cryptographic hash function
+modeled as a random oracle or a
+[pseudorandom function](https://en.wikipedia.org/wiki/Pseudorandom_function_family).
+There are numerous options available for one or both models:
+SHA, BLAKE, AES, ChaCha, etc.
+
+In our implementation, we will let the user provide his own function.
+Hash functions evolve, old ones get broken (SHA-1), and so our ALBA
+implementation shouldn't get stuck with an inefficient / broken hash function.
+Moreover, the user might want to instantiate ALBA with a PRF (e.g. AES) instead
+of a hash function modeled as a random oracle for better theoretical guarantees.
+With Rust's advanced trait system, making the hash function generic should be
+easy, and the runtime code performance should be fast due to compile time
+function resolution.
+
+We will provide an example implementing a hash function with the right interface
+for ALBA using, for example, BLAKE underneath, but with a warning that
+it is provided for demonstration / prototyping only and should not be used in
+production.
+
+One reason is that when ALBA is used alongside other sub-protocols that also
+need hashing, it is the user's responsibility to implement
+[domain separation](https://en.wikipedia.org/wiki/Domain_separation)
+to provide the different
+sub-protocols with independent hash functions.
+Otherwise, the underlying assumptions needed for security might break and the
+larger protocol can become insecure
+(see [here](https://link.springer.com/chapter/10.1007/978-3-030-45724-2_1)).
+
+Additionally, different hash functions (e.g., SHA and BLAKE) shouldn't mix and
+match.
+The following example is contrived, but it demonstrates why using different
+hash function constructions within the same protocol can be insecure.
+Suppose $H$ and $H'$ are two distinct random looking hash functions.
+You might expect that the function $G$ given by $G(x) = H(x) \oplus H'(x)$,
+where $\oplus$ is the bitwise xor, is also random looking, but it's not true.
+Let $H$ be a hash function modeled as a random oracle and define
+$H(x) = \sim H'(x)$,
+where $\sim$ is bitwise not; then $H'$ is also random looking
+but $G(x) = 111...111$ for all $x$.
+Even if we use domain separation for $H$ and $H'$, we cannot guarantee that
+$G(x) = H(0, x) \oplus H'(1, x)$ is random looking.
+Define $H'(b, x) = H(\sim b, x)$; then $G(x) = 000...000$ for all $x$.
+
+## Generic data elements
```

Comment:
I am not sure I fully follow this paragraph and wonder if it belongs to here (or if it should be rephrased). Why not simply say that data elements are completely generic as long as you can test equality on them?

Reply:
And that they are hashable